### ISSUE TITLE
Golang 1.5+ vendoring and middlewares deps

### ISSUE THREAD

**Issue by agalitsyn** [vulcand/vulcand#281]
Before GO15VENDOREXPERIMENT vendor deps was used  by external middlewares

For now this is restricted by go, for example we can see message like 

```
imports github.com/vulcand/vulcand/vendor/github.com/codegangsta/cli: use of vendored package not allowed
```

Without direct vendoring  we have broken deps

```
../../agalitsyn/vulcand-tmirror/tmirror/tmirror.go:20: cannot use FromCli (type func(*"github.com/codegangsta/cli".Context) (plugin.Middleware, error)) as type plugin.CliReader in field value
../../agalitsyn/vulcand-tmirror/tmirror/tmirror.go:21: cannot use CliFlags() (type []"github.com/codegangsta/cli".Flag) as type []"github.com/vulcand/vulcand/vendor/github.com/codegangsta/cli".Flag in field value
```

Explanation of problem https://github.com/mattfarina/golang-broken-vendor

So I tried to remove this dep in https://github.com/vulcand/vulcand/pull/280 but it's wrong

We should use more advanced dependency tool like https://github.com/Masterminds/glide or move to ideology of absolutely independent build of middlewares, for example use https://github.com/natefinch/pie

**Comment by agalitsyn** [comment]
So, anyone have idea? For now I can't compile vulcan with any external middleware budled with vbundle

**Comment by tboerger** [comment]
The vendoring will enabled already on 1.6 by default, 1.7 will even remove the deactivation of vendoring. So this needs to be fixed differently.

The best should be to build real plugins, e.g. pie or https://github.com/hashicorp/go-plugin

From my side I don't have a strong opinion on that as I'm not using any plugin for vulcand.

**Comment by agalitsyn** [comment]
@klizhentas please help us with your vision on this problem. Actually now vulcan is degraded by custom middleware feature compare to previous versions

**Comment by tboerger** [comment]
What about simply using proper import paths?

**Comment by klizhentas** [comment]
@agalitsyn sure, I will investigate

**Comment by Nesze** [comment]
Hi,

Is there any update on this? (even any hackish solution to bypass the problem?)
We face the same issue as we implement our own Auth middleware so it is quite annoying.
Let me know if there is anything we can do.

**Comment by martingartonft** [comment]
I've looked at this and I can't see any options other than the following:
1) Un-vendor codegangsta/cli
2) Don't expose any codegangsta/cli types in from the functions in github.com/vulcand/vulcand/plugin. (doing (2) may mean replicating quite a bit of the codegangsta/cli interfaces, most likely as interfaces)

Am I wrong?

It not, which is the preferred solution?  I'm willing to help if I can.

**Comment by agalitsyn** [comment]
> 1) Un-vendor codegangsta/cli

Vendoring is good for safety of working state for project. External middlewares and vulcand will always have the same version of codegangsta/cli, because it's bundled in one binary from GOPATH. But from build to build version will change and without vendoring we don't have ability to control it.

Also it's will not solve the problem, because I want to use oxy lib in my middleware, as well as vulcand core plugins.

**Comment by agalitsyn** [comment]
Also I tried to hack prototype with RPC-based plugin system like go-plugin or pie. IMO it looks like inefficient way for vulcand plugins, because it's just a chain of http handlers on one request.

**Comment by martingartonft** [comment]
I agree about vendoring.

I don't see how/why "External middlewares and vulcand will always have the same version of codegangsta/cli" though, because vulcand has a vendored version, and according to the go vendoring rules the external middlewares will not be able to see that at build time, so they will have to have their own version, but then the types won't match at compile time (which is the exact problem as I'm seeing it right now when I try)

A also felt the RPC approach wasn't suitable.

So I still don't see a workable solution other than the two I listed.  (2) would be my preference, but it means duplicating a subset of the codegangsta/cli API surface.  I just want to hear what others think of that solution, or if there is another alternative.

**Comment by martingartonft** [comment]
Did anyone have a better suggestion than (2) above?  I don't want to work on this if people think it's insane.

**Comment by wrotki** [comment]
If anyone still needs a workaround:

Applying the vendoring procedure (I used glide) to copy dependencies to the vendor directory creates two copies of codegangsta/cli:

- <project root>vendor/github.com/vulcand/vulcand/vendor/github.com/codegangsta/cli 
- <project root>vendor/github.com/codegangsta/cli 

The compiler complains about functions using Context and other types defined at <project root>vendor/github.com/vulcand/vulcand/plugin/middleware.go, indicating that these type definitions use the first path. I don't quite get it, as according to the Golang 1.5+ vendoring experiment rules as I understand them it should be using the second one . 

But the project builds fine after I simply delete the first directory, which eliminates type resolution ambiguity.

**Comment by wrotki** [comment]
And now I think running 'glide update -s -v -u' does this automatically

**Comment by xgt001** [comment]
@wrotki  thank you, this worked

**Comment by gchaincl** [comment]
@xgt001 I don't get it, how did you make it work by running glide?

**Comment by xgt001** [comment]
I tried @wrotki's earlier solution of deleting one of the redundant codegangsta libraries in vulcand

**Comment by archisgore** [comment]
@ghcaincl, it's the -v option in glide:

--strip-vendor, -v		Removes nested vendor and Godeps/_workspace directories. Requires --strip-vcs

Basically it'll take out sub-nested vendored packages inside any packages you vendor yourself. It ends up ensuring only one copy of the package remains.

**Comment by horkhe** [comment]
Yep, you need to flatten all dependencies at the tope level vendor folder. Please check out the docs of your vendoring tool.

**Comment by designgrill** [comment]
An ELI5 version for someone needing quick help.

Golang 1.5 onwards natively supports vendoring with `vendor/` in the project root given the first preference instead of go root. However as vulcand itself has vendored dependencies, go build system doesn't like it.

One way to resolve this issue is to issue the following commands in your project root:
```
# Clone the main vulcand repo
git clone --depth 1 https://github.com/vulcand/vulcand vendor/github.com/vulcand/vulcand

# Move the codegangsta/cli dependency out in its own folder
# Note that the original codegangsta/cli has move to a different repo
mv vendor/github.com/vulcand/vulcand/vendor/github.com/codegangsta vendor/github.com/
```
